### ISSUE TITLE
why crash?

### ISSUE THREAD

**Issue by sepcity** [yyzybb537/libgo#216]

**Comment by SunnyChenkang** [comment]
这段代码没有问题的，我自己也测试下也没问题；

**Comment by wangqiangnpq** [comment]
~LockedChannelImpl() {
        DebugPrint(dbg_mask_ & dbg_channel, "[id=%ld] Channel destory.", this->getId());

        assert(lock_.try_lock());
    }

是不是这段try_lock()会崩溃 ？ 你是怎么解决的

**Comment by yyzybb537** [comment]
崩溃说明没用对，还有地方在使用就析构了

**Comment by wangqiangnpq** [comment]
我跑的是工程里面sample3的例子，确实main函数直接走完就析构了

**Comment by wangqiangnpq** [comment]
virtual void Excute() 
	{
		co_chan<int> ch_0;

		go[=]{
			ch_0 << 1;
		};
		go[=]{
			int i;
			ch_0 >> i;
			printf("i = %d\n", i);
		};
		

		std::thread([=] { co_sleep(2000); co_sched.Stop();}).detach();
		co_sched.Start();
		
	}
 就是这样, 下面的1已经打印出来了，说明chan已经使用过了，然后被析构还有什么地方lock过呢？

**Comment by wangqiangnpq** [comment]
如果我想要把异步的东西当作同步在调用，我得申请chan并且管理它么，因为我没看到 yield对应的resume接口在哪。

**Comment by Invisible-Boy** [comment]
> ```
> ~LockedChannelImpl() {
>     DebugPrint(dbg_mask_ & dbg_channel, "[id=%ld] Channel destory.", this->getId());
> 
>     assert(lock_.try_lock());
> }
> ```
> 
> 是不是这段try_lock()会崩溃 ？ 你是怎么解决的

这个问题是因为在debug模式下assert语句会执行。在析构函数中对lock_加锁成功以后，析构函数正常进行，最后会把成员变量lock_销毁，但是此时lock_是上锁状态，导致throw 异常。把这句注释掉就可以了。
